server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -377,13 +377,13 @@ private void balance() {
                 logger.trace("skipping rebalance as it is disabled");
                 return;
             }
-            if (nodes.size() < 2) { /* skip if we only have one node */
-                logger.trace("skipping rebalance as single node only");
-                return;
-            }
 
             // Balance each partition
             for (NodeSorter nodeSorter : nodeSorters) {
+                if (nodeSorter.modelNodes.length < 2) { /* skip if we only have one node */
+                    logger.trace("skipping rebalance as the partition has single node only");
+                    continue;
+                }
```

Comment:
This change means we can skip rebalancing for one partition but not the other. I think that's OK. But please let me know if there are subtle connections between the two balancing rounds and must run in tandem.